odl/solvers/nonsmooth/difference_convex.py

```diff
@@ -0,0 +1,243 @@
+        g.convex_conj.gradient(h.gradient(x), out=x)
+
+        if callback is not None:
+            callback(x)
+
+
+def prox_dca(x, g, h, niter, gamma, callback=None):
+    """Proximal DCA of Sun, Sampaio and Candido.
+
+    This algorithm solves a problem of the form ::
+
+        min_x g(x) - h(x)
+
+    where ``g`` and ``h`` are two proper, convex and lower semicontinuous
+    functions.
+
+    Parameters
+    ----------
+    x : `LinearSpaceElement`
+        Initial point, updated in-place.
+    g : `Functional`
+        Convex part. Needs to provide a `Functional.convex_conj` with a
+        `Functional.proximal` factory.
+    h : `Functional`
+        The negative of the concave part. Needs to provide a
+        `Functional.proximal` factory.
+    niter : int
+        Number of iterations.
+    gamma : positive float
+        Stepsize of the proximal steps.
+    callback : callable, optional
+        Function called with the current iterate after each iteration.
+
+    Notes
+    -----
+    The algorithm was proposed as Algorithm 2.3 in
+    `[SSC2003] <http://www.global-sci.org/jcm/readabs.php?vol=21&\
+no=4&page=451&year=2003&ppage=462>`_. It solves the problem
+
+    .. math ::
+        \\min g(x) - h(x)
+
+    by involving subgradients of :math:`h` and proximal points of :math:`g^*`.
+    The iteration is given by
+
+    .. math ::
+        y_n \\in \\partial h(x_n), \\qquad x_{n+1}
+            = \\mathrm{Prox}^{\\gamma}_g(x_n + \\gamma y_n)
+
+    Here, :math:`\\gamma` is the stepsize parameter ``gamma``.
+
+    References
+    ----------
+    [SSC2003] Sun, W, Sampaio R J B, and Candido M A B. *Proximal point
+    algorithm for minimization of DC function*. Journal of Computational
+    Mathematics, 21.4 (2003), pp 451--462.
+
+    [TA1997] Tao, P D, and An, L T H. *Convex analysis approach to d.c.
+    programming: Theory, algorithms and applications*. Acta Mathematica
+    Vietnamica, 22.1 (1997), pp 289--355.
+    """
+    space = g.domain
+    if h.domain != space:
+        raise ValueError('`g.domain` and `h.domain` need to be equal, but '
+                         '{} != {}'.format(space, h.domain))
+    for _ in range(niter):
+        g.proximal(gamma)(x + gamma * h.gradient(x), out=x)
+
+        if callback is not None:
+            callback(x)
+
+
+def doubleprox_dc(x, y, g, h, phi, K, niter, gamma, mu, callback=None):
+    r"""Double-proxmial gradient d.c. algorithm of Banert and Bot.
+
+    This algorithm solves a problem of the form::
+
+        min_x g(x) + phi(x) - h(Kx).
+
+    Parameters
+    ----------
+    x : `LinearSpaceElement`
+        Initial primal guess, updated in-place.
+    y : `LinearSpaceElement`
+        Initial dual guess, updated in-place.
+    g : `Functional`
+        The nonsmooth part of the convex part of the problem. Needs to provide
+        a `Functional.proximal` factory.
+    h : `Functional`
+        The functional involved in the concave part of the problem. Needs to
```

Comment:
Harmonize wording with the other methods

Reply:
The problem is that I am describing things which are not the same. In the doubleprox method, the concave part has the form –(h ∘ K) with an operator K, whereas in the other methods it has just the form –h (and it means no loss of generality because K* s K is a subgradient of h ∘ K if s is a subgradient of h). Therefore, in my opinion, the wordings should be different.